### PR TITLE
Bitmap encoder writes V3 header as default

### PR DESCRIPTION
**Pull request by brianpopow**
### Prerequisites

- [x] I have written a descriptive pull-request title
- [x] I have verified that there are no overlapping [pull-requests](https://github.com/SixLabors/ImageSharp/pulls) open
- [x] I have verified that I am following matches the existing coding patterns and practice as demonstrated in the repository. These follow strict Stylecop rules :cop:.
- [x] I have provided test coverage for my change (where applicable)

### Description

Some bitmap decoders seem to have problems with Bitmap V4 header in combination with 24 bits per pixel and the BITFIELDS compression type (see: #886).

This PR changes the Bitmap encoder to write a V3 header as default. This should be supported by any decoders.
Introduces a new encoder option `SupportTransparency`: With this option a V4 header will be written with BITFIELDS compression.

### CODE REVIEW

File: tests/ImageSharp.Tests/Formats/Bmp/BmpEncoderTests.cs [hunk 1]
```diff
@@ -102,15 +104,43 @@ public void Encode_IsNotBoundToSinglePixelType<TPixel>(TestImageProvider<TPixel>
...
             where TPixel : struct, IPixel<TPixel> => TestBmpEncoderCore(provider, bitsPerPixel);
 
-        private static void TestBmpEncoderCore<TPixel>(TestImageProvider<TPixel> provider, BmpBitsPerPixel bitsPerPixel)
+        [Theory]
+        [WithFile(Bit32Rgb, PixelTypes.Rgba32 | PixelTypes.Rgb24, BmpBitsPerPixel.Pixel32)]
+        [WithFile(Bit32Rgba, PixelTypes.Rgba32 | PixelTypes.Rgb24, BmpBitsPerPixel.Pixel32)]
+        [WithFile(WinBmpv4, PixelTypes.Rgba32 | PixelTypes.Rgb24, BmpBitsPerPixel.Pixel32)]
+        [WithFile(WinBmpv5, PixelTypes.Rgba32 | PixelTypes.Rgb24, BmpBitsPerPixel.Pixel32)]
+        // WinBmpv3 is a 24 bits per pixel image
+        [WithFile(WinBmpv3, PixelTypes.Rgb24, BmpBitsPerPixel.Pixel24)]
```
Comment (antonfirsov) [inline tests/ImageSharp.Tests/Formats/Bmp/BmpEncoderTests.cs]: 👍 these tests are great! 💯

**Comment by brianpopow** [comment]
I just wanted to clarify that rejecting bitmaps with BITFIELDS and 24 bpp is a valid thing to do. While re-reading the bitmap documentation again while i was making the changes for this PR, BITFIELDS should only be used with 16 and 32 bits per pixel. 
It was a mistake, that i have allowed encoding 24 bpp with BITFIELDS.

**Comment by JimBobSquarePants** [comment]
I'll get this reviewed asap

File: src/ImageSharp/Formats/Bmp/BmpEncoder.cs [hunk 2]
```diff
@@ -18,6 +18,11 @@ public sealed class BmpEncoder : IImageEncoder, IBmpEncoderOptions
         /// </summary>
         public BmpBitsPerPixel? BitsPerPixel { get; set; }
 
+        /// <summary>
+        /// Gets or sets a value indicating whether the encoder should support transparency.
+        /// </summary>
+        public bool SupportTransparency { get; set; }
```
Comment (JimBobSquarePants) [inline src/ImageSharp/Formats/Bmp/BmpEncoder.cs]: Wouldn't `BmpBitsPerPixel.Pixel32` implicitly mean this?

**Comment by brianpopow** [comment]
@JimBobSquarePants 
> Wouldn't BmpBitsPerPixel.Pixel32 implicitly mean this?

It is complicated. With a V3 header and 32 bits per pixel (and compression type `no compression`), the 4th byte is just considered a padding byte not actually a alpha channel. This is how a V3 palette entry looks like:

```
typedef struct _Win3xPaletteElement
{
	BYTE Blue;      /* Blue component */
	BYTE Green;     /* Green component */
	BYTE Red;       /* Red component */
	BYTE Reserved;  /* Padding (always 0) */
} WIN3XPALETTEELEMENT;
````

My usual reference source is: [fileformat.info](http://www.fileformat.info/format/bmp/egff.htm)
Some decoders (imagesharp and mozilla included) try to guess if there is an alpha channel present: If there is a none zero 4th byte present, it has an alpha component (see issue #732).

I think the correct way to handle transparency is with the BITFIELDS compression type. 

If we would follow the documentation strictly, we should make the image opaque when we write a V3 header. This is currently not the case. It would not hurt any decoder, because most of them probably ignore the 4th byte rightfully anyway. What is your opinion on this, should we make it opaque with V3 headers?

I could try to expand the comment on `SupportTransparency` to make this more clear. 

My first thought when i was implementing this was add an option which header to write (V3 or V4) and
another option compression type: (none or BITFIELDS), but which end user would actually understand the implication of those anyway. Also this adds another level of complexity, for example V3 + BITFIELDS does not mean transparency is supported, that would be the compression type: BI_ALPHABITFIELDS (and BI_ALPHABITFIELDS does not make much sense in combination with V4 header).

**Comment by JimBobSquarePants** [comment]
@brianpopow Thanks for the thorough explanation. That all makes sense. 

Let's flesh out the comment on the `SupportTransparency` property then dictating exactly what setting are implemented when the property is set.

**Comment by brianpopow** [comment]
I have changed the comment on the `SupportTransparency` option with the implication to the encoder.

Im still on the fence whether or not we should set the alpha / reserved byte to 0, if we write a V3 header with 32 bbp. I have not changed the behavior of the encoder behavior in that regard. If for example a png with an alpha channel is loaded and the again saved as bmp with a V3 header, the encoder would write the alpha channel loaded from the original image. 
All decoders i know of can decode those images, but only some (e.g. gimp, imagemagick, imagesharp) will look for none zero values in the alpha channel to determine if there is transparency present. Others will decode it as full opaque, which is also correct behavior according to the documentation for V3 header bitmaps.

File: src/ImageSharp/Formats/Bmp/BmpEncoder.cs [hunk 3]
```diff
@@ -18,6 +18,14 @@ public sealed class BmpEncoder : IImageEncoder, IBmpEncoderOptions
         /// </summary>
         public BmpBitsPerPixel? BitsPerPixel { get; set; }
 
+        /// <summary>
+        /// Gets or sets a value indicating whether the encoder should support transparency.
+        /// Note: Transparency support only works together with 32 bits per pixel. This option will
+        /// change the default behavior of the encoder of writing a bitmap version 3 info header with no compression.
+        /// Instead a bitmap version 4 info header will be written with the BITFIELDS compression.
```
Comment (JimBobSquarePants) [inline src/ImageSharp/Formats/Bmp/BmpEncoder.cs]: Perfect!

**Pull request merged**